synapse/handlers/room.py

```
@@ -1217,6 +1218,114 @@ async def filter_evts(events: List[EventBase]) -> List[EventBase]:
+            likely_domains = [
+                domain for domain, depth in curr_domains if domain != self.server_name
+            ]
+
+            # Loop through each homeserver candidate until we get a succesful response
+            for domain in likely_domains:
+                try:
+                    remote_response = await self.transport_layer.timestamp_to_event(
+                        domain, room_id, timestamp, direction
+                    )
+                    logger.debug(
+                        "get_event_for_timestamp: response from domain(%s)=%s",
+                        domain,
+                        remote_response,
+                    )
+                    if not remote_response:
+                        continue
+
+                    remote_event_id = remote_response.get("event_id", None)
+                    if remote_event_id:
+                        # TODO: Do we want to persist this as an extremity?
+                        # TODO: I think ideally, we would try to backfill from
+                        # this event and run this whole
+                        # `get_event_for_timestamp` function again to make sure
+                        # they didn't give us an event from their gappy history.
+                        return remote_event_id
+
+                    continue
+                except Exception as ex:
+                    logger.exception(
+                        "Failed to fetch /timestamp_to_event from %s because of exception(%s) %s args=%s",
+                        domain,
+                        type(ex).__name__,
+                        ex,
+                        ex.args,
+                    )
+                    continue
+
+        if not local_event_id:
+            raise SynapseError(
+                404,
+                "Unable to find event from %s in direction %s" % (timestamp, direction),
+                errcode=Codes.NOT_FOUND,
+            )
```

The return type of `Optional[str]` may be misleading because we raise instead of returning `None`